Add a simple pytest coverage check on workflow_dispatch or merge queue

zaneselvans opened the pull request:
# Overview

Successor to the ill-fated #3352 which snuck through the merge queue when it was supposed to fail the coverage check.

- Uses a simple fixed threshold coverage check.
- Tries to trigger `ci-integration` and `ci-coverage` on both `workflow_dispatch` and in the merge queue, but not in a pull request

Closes #3337
Closes #3370 

```[tasklist]
# To-do list
- [x] Try running `ci-integration` with a `workflow_dispatch` trigger and get it to run the tests + coverage.
- [x] Review the PR yourself and call out any questions or issues you have
- [x] Try submitting to merge queue with 100% coverage requirement and get rejected.
- [x] Reset target coverage threshold to... 92.8% (currently [CodeCov says we're at 92.95%](https://app.codecov.io/gh/catalyst-cooperative/pudl))
```

zaneselvans commented on .github/workflows/pytest.yml:
```diff
@@ -50,11 +50,11 @@ jobs:
         with:
           name: coverage-docs
-          path: coverage.xml
+          path: ./*coverage*
```
For all the coverage generating jobs, upload both the XML and DB versions of the coverage data.

zaneselvans commented on .github/workflows/pytest.yml:
```diff
@@ -97,17 +97,17 @@ jobs:
       group: large-runner-group
       labels: ubuntu-22.04-4core
-    if: github.event.merge_group || github.event.workflow_dispatch
+    if: github.event_name == 'workflow_dispatch' || github.event.merge_group
```
Run integration tests (and thus coverage check) only on manual `workflow_dispatch` or if we are in the merge queue.

jdangerx replied:
Huh. Weird quirk of GHA I guess 👍

zaneselvans commented on .github/workflows/pytest.yml:
```diff
@@ -198,11 +198,24 @@ jobs:
       - name: List downloaded files
         run: |
-          ls -R
+          find coverage -type f
```
More focused output -- only the coverage artifacts.

zaneselvans commented on .github/workflows/pytest.yml:
```diff
@@ -198,11 +198,24 @@ jobs:
         uses: codecov/codecov-action@v4
         with:
           directory: coverage
+      - name: Install Micromamba
+        uses: mamba-org/setup-micromamba@v1
+        with:
+          init-shell: bash
+          environment-name: coverage
+          create-args: >-
+            python=3.11
+            coverage>=7.4.1
+      - name: Combine coverage data and check that we have required coverage
+        run: |
+          micromamba run -n coverage coverage combine coverage/*/.coverage
+          micromamba run -n coverage coverage xml
+          micromamba run -n coverage coverage report --fail-under=100
```
Seems like the simplest possible check -- just a fixed percentage threshold.

Could improve by using the `codecov` CLI to track whatever the ongoing coverage is, but this will at least ensure we don't backslide, and CodeCov will keep accumulating data.

jdangerx replied:
OK let's see if this thing fails! And then we can crank the # down.

jdangerx reviewed (approved):
Seems safe enough to try, thanks for poking at this!

jdangerx commented:
```
25 empty files skipped.
Coverage failure: total of 92.8 is less than fail-under=100.0
Error: Process completed with exit code 2.
```
🎉

zaneselvans commented:
- I've cleaned up the coverage configuration to specify the minimum required coverage in `pyproject.toml` so it'll fail with the `make pytest-coverage` command as well, and we only have to update it in one place.
- Also updated the PR template and documentation to clarify what tests folks should run under what circumstances, indicating that `make pytest-coverage` is a good indicator of what will get run in the merge queue.

Also apparently I have broken the unit tests soooo, wtf.  Oh right it makes a coverage report.

zaneselvans commented on .github/workflows/pytest.yml:
```diff
@@ -198,36 +198,22 @@ jobs:
         with:
           directory: coverage
-
-  ci-notify:
-    runs-on: ubuntu-latest
-    if: ${{ always() }}
```
This notification has been uninformative for a long time and creates a bunch of Slack noise. I approved its removal in once of Jan's PRs and have finally gotten fed up enough with it to cut it here...
